0020-governance-delegator-votes.md

```diff
@@ -0,0 +1,134 @@
+# ADR 0020: Governance Support Delegators Votes
+
+## Component
+
+Oasis Core
+
+## Changelog
+
+- 2022-11-02: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+With the current governance voting mechanism ([ADR-0004]), only the active
+validator set is participating in voting. This means that the validators are
+voting on behalf of all their delegators. This ADR proposes a change so that
+each delegator is able to vote with its own stake. The delegators vote acts as
+an override of the validator vote.
+
+## Decision
+
+### Casting Votes
+
+In the current implementation the submitter of a vote needs to be a part of the
+active validator committee at the time the vote is cast. This requirement is
+relaxed so that additionally anyone with a delegation to an active validator
+committee entity can vote.
+
+This change requires an efficient `staking.DelegationsFor` query to obtain a
+list of accounts the submitter is delegating to. Staking state is updated with:
+
+```go
+
+// delegationKeyReverseFmt is the key format used for reverse mapping of
+// delegations (delegator address, escrow address).
+//
+// Value is CBOR-serialized delegation.
+delegationKeyReverseFmt = keyformat.New(0x5A, &staking.Address{}, &staking.Address{})
+```
+
+`state.SetDelegation` function is updated to store both `delegationKeyFmt` and
+the reverse `delegationKeyReverseFmt`.
+`DelegationsFor` query function is updated to use the added reverse mapping.
+
+For completeness the same can be done for debonding delegations, although not
+necessary for the governance changes proposed in this ADR.
+
+#### Alternatives
+
+Possible alternatives that would avoid adding the reverse mapping are:
+
+- Querying `DelegationsTo` for each validator. This results in `num_validators`
+queries per cast vote transaction which is still too much.
+- Allowing anyone to cast votes. Potentially a viable solution, but this could
+result in the number of voters growing uncontrollably large. This might be ok,
+as the proposed vote tallying procedure would ignore those votes. However the
+votes state could still grow problematically big.
+
+### Vote tallying
+
+When a proposal closes, the vote tallying procedure changes to:
+
+```
+1 Precompute stakes for all delegators to validators from shares.
+2 For each validator
+3   For each delegator to the validator
+4     IF validator and delegators votes match (or delegator didn't vote)
+5       Add delegator stake to the validator's vote (or nothing if validator didn't vote)
+6     IF validator and delegator vote don't match
+7       Add delegator stake to the delegator's vote (or nothing if delegator didn't vote)
+```
+
+The voting procedure now iterates over all delegators of the active validator set. The amount of
+work is somewhat predictable as it doesn't depend on the number of voters but on the
+delegators-to-validator sets.
+
+#### Alternatives
```

Comment:
Would it be possible to have an "iterative" alternative, where tallying is performed when the vote is cast in order to amortize the cost? The downside is the possibly greatly increased complexity as now one would need to also handle delegations/undelegations after the vote. So this is something I would only consider in case the cost of the above is too great with a reasonable amount of delegators (e.g. 100,000).

Would also be good to look at what other governance implementations are doing with regards to tallying governance votes.

Reply:
> Would it be possible to have an "iterative" alternative...

I believe that the greatly increased complexity is likely not worth it - specially since it would be really hard to reason about the results, you'd need to go over the whole voting period and doublecheck all the updates. But will keep in mind in case the alternatives turn out to be too expensive. 

> Would also be good to look at what other governance implementations are doing with regards to tallying governance votes.

Cosmos-sdk does something similar to the alternative approach proposed here. They iterate over voters, their delegations and validators.  See: https://github.com/cosmos/cosmos-sdk/blob/dc004c85f2e8b8fb4f66caac2703228c5bf544cf/x/gov/keeper/tally.go#L37-L90

They also limit voting similarly as described in our proposal.

(will check some other implementations later)

---

Thinking about it a bit more:
IF vote casting is implemented as currently described (require voters to be validators or delegators to validators)
THEN this alternative approach makes more sense since the number of voters will be at most the number of delegators (and likely much less) (unless extreme pathological edge cases where validator set fluctuates greatly).
So I think that iterating over the voters (and their delegators) does actually make more sense than going over the whole validator-delegator set.

Reply:
> So I think that iterating over the voters (and their delegations) does actually make more sense than going over the whole validator-delegator set.

Agreed.